### DISCUSSION TITLE
apple silicon: how to switch back to gcc-10

### DISCUSSION THREAD

**Discussion by scivision** [Homebrew/discussions#1443] [Everyday usage] [answered]
The [gcc formula](https://formulae.brew.sh/formula/gcc@10) shows it's possible to go back to gcc-10 on Intel, but not for M1. We are experiencing a serious GCC-11 Fortran regression (a problem with GCC 11 itself, not Homebrew), and our M1 users would like to switch back to GCC-10. 

However, the GCC formula page doesn't show a way for M1 to use `gcc@10`. Is there a way to enable this fallback?

**Comment by carlocab** [comment, marked as answer]
You'll need to maintain this in your own tap. Try
```
brew extract --version=10.2.0 gcc username/tap-name
```
You'll need to create a new tap for this. See `brew help tap-new` for details.

**Reply by scivision** [reply]
Thank you Carlo, this is the complete workaround that enabled us to use GCC 10 on the M1 until GCC 11 is fixed:

https://github.com/scivision/homebrew-tap#gcc-10

**Reply by carlocab** [reply]
No problem. Good to know you got that set up without any problems.

**Reply by carlocab** [reply]
By the way, have you reported the regressions in `gfortran` upstream? Upstream here might actually be https://github.com/iains/gcc-darwin-arm64

**Reply by scivision** [reply]
I don't think it's arch-specific. It happens in a large Fortran program that has numerous CI tests across compiler versions and operating systems. I am working on creating a minimal case to reproduce for upstream GCC but that might take me some time.

**Reply by carlocab** [reply]
CC @fxcoudert, as you might be interested

**Reply by fxcoudert** [reply]
Definitely interested in knowing what the problem is. Note that GCC 10 on the M1 was an experimental backport of the ARM64-Darwin support, and is not actively supported or maintained anymore.

**Reply by carlocab** [reply]
@scivision You'll want to give FX here even a brief summary of the regression. They'll definitely be able to help you.

**Reply by scivision** [reply]
I have a library [h5fortran](https://github.com/geospace-code/h5fortran) that is an object-oriented interface for HDF5 from Fortran. I use types with Generic procedure to handle the polymorphism with an object-oriented interface. This is where the problem happens, a runtime segfault, if I call it like this psuedocode

```fortran
real(real64) :: x
call obj%mypoly(real(x))  !< runtime segfault GCC 11 only

real(real32) :: y
y = real(x)
call obj%mypoly(y)  !< OK GCC 11
```

In this trivial example, mypoly's dummy argument is `intent(in)`.  I choose to coerce to real32 to save disk space with HDF5 files.

I tried to make a simple failing example within h5fortran but wasn't successful in making a failure yet.  I have another program [Gemini3D](https://github.com/gemini3d/gemini3d) that is a suite of Fortran and C programs--that's where the runtime failure occurs. So it seems when I get a big stack of types and polymorphism that the problem happens. I didn't want to bother GCC devs with installing Gemini3D.  However if they're interested I can give a very simple procedure for that--Gemini3D build all prereqs automatically via CMake like it's own package manager.

My first guess would be that Gfortran 11 may be treating dummy argument with `intent(in)` as `intent(inout)` maybe.
